### ISSUE TITLE
Heretic demo h5f3-803.lmp desyncs

### ISSUE THREAD

**Issue by chocolate-import** [chocolate-doom/chocolate-doom#299]
> The following bug was [originally reported](https://sourceforge.net/p/chocolate-doom/bugs/299/) on Sourceforge by [fraggle](https://sourceforge.net/u/fraggle/profile/), 2013-10-30 02:36:30.703000:

Bug report from marineController (Doomworld):

  The demo h5f3-803.zip desyncs at approx. 6m 30s.

http://lsvc.free.fr/Heretic-N/Demos/Max-bp/

**Comment by njankowski** [comment]
Building from master 37c1162, I could not get this to desync. The player seems to reach a secret exit just fine. The text file inside is also incorrect. This is E5M3 Quay.

Here is the old Doomworld thread on the issue:
https://www.doomworld.com/vb/source-ports/65863-chocolate-doom-version-2-beta
Link to demo is dead. It can be found here:
http://82.66.133.154:85/Heretic-N/Demos/Max-bp/

**Comment by jmtd** [comment]
Desyncs for me with current git HEAD

**Comment by njankowski** [comment]
Built 5e162ed and had no issue. Only used -iwad and -playdemo arguments.
Demo SHA1:
279deb5df99c93e16bb84ead7060629505503e8c
Heretic (1.3) WAD SHA1:
f489d479371df32f6d280a0cb23b59a35ba2b833

**Comment by chungy** [comment]
Doesn't desync for me either. I'm on Linux x86_64, built with GCC 5.3.0

**Comment by jmtd** [comment]
Interesting. My sums match @njankowski . I built on OS X so that's llvm

```
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin15.3.0
Thread model: posix
```

**Comment by njankowski** [comment]
Using Debian Jessie x86_64, identical results to @chungy except I was using GCC 4.9.2 and Clang 3.5.

**Comment by njankowski** [comment]
Desync starts at game tick 4683.

GCC compiled version has three more calls to P_Random during game tick 4683. These three calls occur in the next game tick with the Clang compiled version.

A screenshot of game tick 4683 for your viewing pleasure:
![htic00](https://cloud.githubusercontent.com/assets/11827629/13017087/7cec88ba-d190-11e5-8121-61731bf4013c.png)

**Comment by CapnClever** [comment]
All this was tested in Win7 64-bit.

I started by checking out the demo under the version it was most likely to be used with at the time of reporting: 2.0.0-beta2. Here I was able to reproduce the desync at the approximated time given by the reporter (which, for reference, is not the same place as referenced by @njankowski).

Tested again under 2.2.1, no desync was found. (Presumably the Win32 binary is compiled via MinGW.)

Tested again on master branch de82ea5, compiled using MSVC2015, desync was found approximately the same place as the previous post.

In conclusion, whatever was broken before probably isn't anymore, but there does appear to be a compiler-dependent desync problem hanging around.

**Comment by jmtd** [comment]
retested with Apple LLVM version 7.3.0 (clang-703.0.31) and c-d de82ea5061aeb1646c0fb6c2e25383c8416ad089, alas bug remains.

**Comment by turol** [comment]
Links are dead.

**Comment by njankowski** [comment]
Found one of the demos on a backup disk.
[h5f3-803.zip](https://github.com/chocolate-doom/chocolate-doom/files/4290611/h5f3-803.zip)
The Doomworld thread also mentions `h5s2-344.zip`, but that's not mentioned in this issue, and I probably don't have it.
For anyone investigating this issue I would re-read the Doomworld thread, because it seems to have some debugging conversation already from the past.

**Comment by turol** [comment]
Unable to reproduce. At least when playing back with timedemo both GCC- (both 5 and 9) and clang (4 and 9) -built binary the demo looked sane, in sync and similar to heretic in dosbox. Also with vanilla heretic timedemo desynced immediately but playdemo worked. Can someone still reproduce this?
